src/lib.rs

```diff
@@ -1875,6 +1875,343 @@ fn digraph_dijkstra_shortest_path_lengths(
     Ok(out_dict.into())
 }
 
+/// Calculate the the shortest length from all nodes in a
+/// :class:`~retworkx.PyDiGraph` object
+///
+/// This function will generate the shortest path from a source node using
+/// Dijkstra's algorithm.
+///
+/// :param graph: The input :class:`~retworkx.PyDiGraph` to use
+/// :param weight_fn: A weight function for an edge. It will accept
+///     a single argument, the edge's weight payload object and will return a
+///     float which will be used to represent the weight/cost of the edge
+///
+/// :return: Dictionary of paths. The keys are destination node indices and
+///     the dict values are dicts of the target node and the length of the
+///     shortest path to that node. For example::
+///
+///         {
+///             0: {1: 2.0, 2: 2.0},
+///             1: {2: 1.0},
+///             2: {0: 1.0},
+///         }
+///
+/// :rtype: dict
+#[pyfunction]
+#[text_signature = "(graph, edge_cost_fn, /)"]
+pub fn digraph_all_pairs_dijkstra_path_lengths(
+    py: Python,
+    graph: &digraph::PyDiGraph,
+    edge_cost_fn: PyObject,
+) -> PyResult<PyObject> {
+    let edge_cost_callable = |a: &PyObject| -> PyResult<f64> {
+        let res = edge_cost_fn.call1(py, (a,))?;
+        let raw = res.to_object(py);
+        raw.extract(py)
+    };
+    let edge_bound: usize =
+        graph.graph.edge_indices().map(|x| x.index()).max().unwrap();
+    let mut edge_weights: Vec<Option<f64>> = Vec::with_capacity(edge_bound);
+    for index in 0..=edge_bound {
+        let raw_weight = graph.graph.edge_weight(EdgeIndex::new(index));
+        match raw_weight {
+            Some(weight) => {
+                edge_weights.push(Some(edge_cost_callable(weight)?))
+            }
+            None => edge_weights.push(None),
+        };
+    }
+    let edge_cost = |e: EdgeIndex| -> PyResult<f64> {
+        match edge_weights[e.index()] {
+            Some(weight) => Ok(weight),
+            None => Err(PyIndexError::new_err("No edge found for index")),
+        }
+    };
+    let node_indices: Vec<NodeIndex> = graph.graph.node_indices().collect();
+    let out_map: Vec<(usize, Vec<(usize, f64)>)> = node_indices
+        .into_par_iter()
+        .map(|x| {
+            let out_vec: Vec<(usize, f64)> =
+                dijkstra::dijkstra(graph, x, None, |e| edge_cost(e.id()), None)
+                    .unwrap()
+                    .iter()
+                    .map(|(index, cost)| (index.index(), *cost))
+                    .collect();
+            (x.index(), out_vec)
+        })
+        .collect();
+    let out_dict = PyDict::new(py);
+    for (index, entries) in out_map {
+        let inner_dict = PyDict::new(py);
+        for (map_index, val) in entries {
+            if map_index == index {
+                continue;
+            }
+            inner_dict.set_item(map_index, val)?;
+        }
+        out_dict.set_item(index, inner_dict)?;
+    }
```

Comment:
Since the dict creation is similar for all instances, in order to not write redundant code, wouldn't be better to abstract this part of the code to be reusable in other instances?

Reply:
This is a good point, and I think that we actually really want to go even one step further and create some custom return types that implement the mapping protocol, that way we don't need the overhead of dict creation. I will do that in a separate PR because it will be a larger scope than this as it will modify the existing dijkstra functions as well.

Reply:
I've done this here: https://github.com/Qiskit/retworkx/pull/287 if this PR merges first, I'll update that to add a new custom return type that wraps those. If that merges first I'll update it here and use new classes here that wrap those.

Reply:
Awesome! I'll take a look on #287.

Reply:
Ok, done in: https://github.com/Qiskit/retworkx/pull/278/commits/4c0478ac942231be54d7d0fbc47abe50693a2c6b and https://github.com/Qiskit/retworkx/pull/278/commits/59145f2935d4c048c59b0c795ef12e5b83e6dc10

This has a very noticeable difference in performance running the benchmark script I did before it goes from ~93 seconds to ~42 seconds locally.